src/librustc_mir/interpret/intern.rs

```diff
@@ -366,6 +369,17 @@ pub fn intern_const_alloc_recursive<M: CompileTimeMachine<'mir, 'tcx>>(
             // dangling pointer
             throw_unsup!(ValidationFailure("encountered dangling pointer in final constant".into()))
         } else if let Some(_) = ecx.tcx.alloc_map.lock().get(alloc_id) {
+            // If we encounter an `AllocId` that points to a mutable `Allocation`,
+            // (directly or via relocations in its `Allocation`), we should panic,
+            // the static rules should prevent this.
```

Comment:
Why is this? *Pointing to* something mutable seems entirely fine even for a constant, as long as the mutable thing isn't *read*. And we already check that anyway.

The point of the FIXME here seems to be to check more than we currently do, but I don't have the slightest clue what it is that would want to check here.

Reply:
> Why is this? Pointing to something mutable seems entirely fine even for a constant, as long as the mutable thing isn't read. And we already check that anyway.

The static rules say we don't have constants pointing to mutable things, so the dynamic checks should mirror that, right?

Reply:
The static checks are whatever they happen to be and I don't know why. Ultimately the *dynamic* checks should be the "source of truth" and the static check justified by approxmating the dynamic checks, not the other way around.

So, "the static checks do that" does nothing to answer my question here: why do we want this restriction? (I also still don't really get what it is that you actually want to check here. What does pointing to mutable things have to do with "string and byte string literals"?)

Reply:
> Ultimately the dynamic checks should be the "source of truth" and the static check justified by approxmating the dynamic checks, not the other way around.

Ok, that makes sense... I'll adjust the comment.

> What does pointing to mutable things have to do with "string and byte string literals"?)

Nothing, this comment is in a code path where immutable allocations also get to, we'd need to check for mutability first, too.